### DISCUSSION TITLE
Able to run in a browser

### DISCUSSION THREAD

**Discussion by ericis** [terrastruct/d2#234] [Feature Requests]
The very first use cases that came to mind for d2 was working within a browser context. However, it's written in Go.

As a result, there are really two options: rewrite in JavaScript or use WebAssembly.

This issue simply requests browser support (not making a network request to a server) to execute d2.

I might personally suggest WebAssembly, since the project is well underway and would take a significant shift otherwise.

**Comment by FFengIll** [comment]
I believe d2 is on the way for it in its own way, pls see #207

**Comment by alixander** [comment]
that's different. that's a playground environment, i believe @ericis is asking about a library callable from client-side browser context. 

Indeed, we have plans to port D2 to WebAssembly.

**Comment by ericis** [comment]
Yes, correct @alixander . The simple use case would be building a diagram via JavaScript in the browser, a more advanced example might be the MermaidJS "Live editor" and the best example I can think of might be an embedded web editor like Eclipse Theia or VS Code complete with a language server for a rich diagram editing experience with live preview.

**Comment by zbysir** [comment]
Have there been any new developments? 

d2 is a great project that can do things that MermaidJS can't, but I think it can run directly in a web browser. Since most documents are static pages, there is no Golang environment.

**Reply by aschrijver** [reply]
WebAssembly is picking up in popularity, both in the browser and on the server (WASI / Component Model). I am also curious about plans in this direction.

**Reply by alixander** [reply]
No new developments as of yet. Perhaps the impetus will be if GitHub wants to integrate rendering D2 natively. Currently there's just too many other priorities that directly help grow the popularity of D2 in more direct ways. Such as features like imports that we've just released. As soon as there's some opportunity like a platform wanting to integrate D2 but can't without Webassembly, we'll port it in a week. But also once D2 gets to 1.0, we'll probably want to line up that release with a Webassembly build.

**Reply by aschrijver** [reply]
Just passing a FYI of [extended Wasm support](https://go.dev/blog/go1.21#a-new-port-to-wasi) in Go 1.21 (WASI Preview 1)

**Reply by PgBiel** [reply]
Hey @aschrijver , thanks for the tip! Can confirm I managed to run one of D2's library usage examples [^1] in `wasmtime` using the steps in this blog post: https://go.dev/blog/wasi
Basically, in the directory with the example's source code, I just had to run (using go 1.21.5) `GOOS=wasip1 GOARCH=wasm go build -o main.wasm` and then `wasmtime --dir "$(realpath .)" main.wasm`, _et voilà_: `out.svg` appeared in the directory.

[^1]: https://github.com/terrastruct/d2/blob/master/docs/examples/lib/1-d2lib/d2lib.go

**Reply by PgBiel** [reply]
Worth mentioning, however, that the resulting binary was 38 MB for that example, which I'd consider to be a bit large for web contexts. Perhaps that can be reduced/optimized, maybe with a bit of help or advice from the D2 team?

For a little bit of context, I'd be interested in seeing a D2 plugin for [Typst](https://typst.app), so that people can easily include D2 diagrams in their documents. Typst supports WASM plugins which can be used when running the Typst compiler either natively or in the browser. Hence why WASM support would be awesome. 🙂

**Reply by alixander** [reply]
It's large because it embeds a javascript runner to run 2 of the layout engines. But it doesn't need to in WASM so part of the work of porting to WASM would be to replace that. https://github.com/terrastruct/d2/issues/366

**Reply by stereobooster** [reply]
> Worth mentioning, however, that the resulting binary was 38 MB for that example

For web yes, but for desktop and SSG (static site generators) it may work. Compiling to WASM would allow to distribute universal binary via npm. Which would result in bigger adoption by JS based SSG, such as Astro, Starlight, VitePress, Next.js (if code executed with server actions), etc.

WDYT? Maybe release as huge WASM (as is) via npm for server usage?

---

This would allow to remove additional step

> You will also need to have D2 installed. You can follow the installation instructions to install D2 on various platforms.
>
> https://astro-d2.vercel.app/getting-started/

> The plugin requires the D2 diagram language to be installed on your system, it is compatible with Linux, Windows and MacOS.
> You can use the following command on Linux to install D2 (official docs).
>
> https://github.com/BadgerHobbs/vitepress-plugin-d2

> Ensure that you have [d2](https://github.com/terrastruct/d2) installed and accessible on your PATH
>
> https://github.com/mech-a/remark-d2

Not sure about VSCode. Would it be possible to run LSP with WASM?

> Note: Requires D2 to be installed on your machine. See https://github.com/terrastruct/d2/tree/master#install for instructions.
>
> https://marketplace.visualstudio.com/items?itemName=terrastruct.d2

**Reply by alixander** [reply]
D2 WASM (and a d2.js package) will be released by end of year.

**Reply by stereobooster** [reply]
I saw that WASM related PR had been merged. Is it possible to create issues or roadmap to show what are the next steps here?

I think it mentioned: JS wrapper, documentation, optimize size... It would be nice to have a list public to track the status. Thanks

**PS** I personally interested only in rendering on the server side (nodejs), nor size, nor LSP feature is big deal to me.

---

https://github.com/terrastruct/d2/issues/2268
https://github.com/terrastruct/d2/issues/2272

**Reply by alixander** [reply]
@stereobooster an alpha is up at https://www.npmjs.com/package/@terrastruct/d2 . We'd benefit from your experience testing it in a node environment and reporting any issues that may come up!

**Reply by stereobooster** [reply]
Hey. Thanks, for releasing it. I will do some experiments. But here is first feedback: is it possible to publish TypeScript definitions for the package?

Package documentation says:

```
compile(input: string, options?: CompileOptions): Promise<CompileResult>
```

But there is no TypeScript definition for `CompileOptions`.

---

**UPD**: also some options are missing. This is what I use currently

```js
  if (options?.theme) args.push(`--theme=${options.theme}`);
  // if (options?.darkTheme) args.push(`--dark-theme=${options.darkTheme}`);
  if (options?.layout) args.push(`--layout=${options.layout}`);
  if (options?.pad !== undefined) args.push(`--pad=${options.pad}`);
  if (options?.scale) args.push(`--scale=${options.scale}`);
  if (options?.sketch) args.push(`--sketch`);
  if (options?.bundle) args.push(`--bundle`);
  if (options?.center) args.push(`--center`);
```

**Reply by alixander** [reply]
thanks
https://github.com/terrastruct/d2/issues/2286
https://github.com/terrastruct/d2/issues/2297

**Reply by stereobooster** [reply]
It works perfectly as far as I can see

**Reply by stereobooster** [reply]
I found one issue. I need to use new instance of each render, otherwise it returns something strange instead of SVG

This doesn't work (when called more than one time):

```ts
const d2Instance = new D2();
export async function d2(
  code: string,
  options?: D2Options
): Promise<string> {
  const result = await d2Instance.compile(code, options);
  return await d2Instance.render(result.diagram, options);
}
```

This works:

```ts
export async function d2(
  code: string,
  options?: D2Options
): Promise<string> {
  const d2Instance = new D2();
  const result = await d2Instance.compile(code, options);
  return await d2Instance.render(result.diagram, options);
}
```

**Reply by alixander** [reply]
What's "something strange"? I tested it and it seems fine: https://github.com/terrastruct/d2/compare/reuse-instance?expand=1#diff-e30335c050d7a53a1e5646b8b866a2a51476b6f55019ae8caa6d2a19744fcad0R34

**Reply by stereobooster** [reply]
I didn't do extended testing (because I found workaround). But I guess you need to use two different diagrams to reproduce error (or maybe not)

```js
const d2 = new D2();
const result1 = await d2.compile("x -> y");
const svg1 = await d2.render(result1.diagram);
const result2 = await d2.compile("a -> b");
const svg2 = await d2.render(result2.diagram);
```

**Reply by alixander** [reply]
What's the error

**Reply by stereobooster** [reply]
If I would have time to reproduce it - I will report it. Maybe it was something unrelated. If you can't reproduce it - high chances it was something else. Just ignore it

Sorry for taking your time

**Reply by alixander** [reply]
Okay, no problem
